README.md
# HADES API

---

## Table of Contents

- [What does it do?](#what-does-it-do)
- [Getting started](#getting-started)
- [How it works](#how-it-works)
    - [Boons](#tickets-table)
    - [Gods](#organisations-table)
- [Status and roadmap](#project-status)
- [FAQ](#faq)

---
## What does it do?

Built for the popular game Hades, the Hades API allows the user to save boons they want to use or sort to a database. These boons will be sorted by their respective gods. This allows for build creation and customization.

---
## Getting started

1. Clone this repository.
2. Navigate to the root folder and boot up the Docker container. For more information regarding Docker, please refer to [their documentation](https://docs.docker.com/).
    ```shell
    docker-compose build
    docker-compose up
    ```
3. To run tests, navigate to the api folder.
    ```shell
    npm test
    ```
4. To see the content in the PostgreSQL database, I recommend using a GUI tool such as TablePlus. You can get it [here](https://tableplus.com/).

---
## How it works

By using the CRUD endpoints, users can access the boons table to add, remove or get boons from the database.

#### Boons


You can modify this table by using the following endpoints:

- `GET /boons`
    - Returns all boons from the database.
- `GET /boon/:uuid`
    - Returns a specific boon by uuid.
- `POST /postboon`
    - Saves a boon to the database.
    - Requires a body with the following properties: 
        ```js
        {
            content: String,
            godid: String
        }
        ```
- `DELETE /deleteboon`
    - Deletes a specific boon by uuid.
    - Requires a body which includes the uuid as a string: 
        ```js
        {
            uuid: "babf5fb0-4d18-11eb-864f-f9e4749272ab"
        }
        ```

## Project Status

This project is in development.

---
## FAQ

**Q: I'm having trouble setting up my development environment, can you help me?**

A: Please refer to [the contribution guidelines](CONTRIBUTING.md) for more information.

**Q: Can I use this project for commercial or private use?**

A: This project falls under the MIT License which allows you to do that. For more information, take a look at [the detailed license](LICENSE.md).


CHANGELOG.md
# Changelog

----

Ver 1.0.1
- added link between boons and gods

Ver 1.0.2
- Fixed the integration tests

Ver 1.0.3
- Added end-to-end testing
- Added more endpoints for gods

Ver 1.0.4
- Added inline documentation
- lost code, added code, now happy once again


hades-api/src/utils/__tests__/app.test.js
const Helpers = require('../helpers');
const app = require("../../server");
const supertest = require("supertest");
const request = supertest(app);

test('Test if UUID is returned', () => {
    expect(Helpers.generateUUID).not.toBeUndefined();
  });


describe("Test if GET /boons", () => {
  it("responds with all boons", async (done) => {
    try {
      const response = await request.get("/boons");
      expect(response.status).toBe(200);
      expect(typeof response.body).toBe("object", done());
    } catch (error) {}
  });
});

describe('Test if POST /boon', () => {
  it('POST /boon', async (done) => {
      const response = await request
          .post('/boon')
          .send({
              godname : "god number 1",
              content : "Adds 50% attack speed"
          });
          expect(response.status).toBe(200, done());
      })           
  });

/*******************************/
/*      Integration test       */
/*******************************/

describe("Test if boon is created and deleted", () => {
  let uuid;
  it("Return 200 when record added to database and return a UUID", async (done) => {
    const response = await request
        .post('/boon')
        .send({
            godname : "god number 1",
            content : "Adds 50% attack speed"
        });
        expect(response.status).toBe(200);
        uuid = response.body.uuid;
        done();
    }) ;
  it("Look for latest record in database", async (done) => {
      const response = await request.get("/boon/" + uuid);
      expect(response.status).toBe(200);
      expect(typeof response.body).toBe("object");
      done();
  });
  it("Return 200 when boon is deleted", async (done) => {
      const response = await request.delete("/boon").send({ uuid: uuid });
      expect(response.status).toBe(200);
      done();
  });
});



hades-api/src/utils/__tests__/endtoend.test.js
const Helpers = require('../helpers');
const app = require("../../server");
const supertest = require("supertest");
const request = supertest(app);

test('Test if UUID is returned', () => {
    expect(Helpers.generateUUID).not.toBeUndefined();
  });

describe("The end-to-end test", () => {
    let boonID;
    let godID;
    it("creates a god", async (done) => {
            const response = await request
                .post("/god")
                .send({
                    name: "Hades",
                    description: "God of the underworld"
                });
            expect(response.status).toBe(200);
            godID = response.body.res[0].uuid;
            done();
    });
    it("gets the newly added god", async(done) => {
            const response = await request
                .get("/god/" + godID)   
            expect(response.status).toBe(200);
            expect(response.body.name).toBe("Hades");
            done();
    })
    it("creates a boon", async (done) => {
            const response = await request
                .post("/boon")
                .send({ 
                    godname : "Hades",
                    content : "Adds 50% attack speed"
            });
            expect(response.status).toBe(200);
            boonID = response.body.uuid;
            done();
    })
    it("gets the newly added boon", async(done) => {
            const response = await request
                .get("/boon/" + boonID)
            expect(response.status).toBe(200);
            expect(response.body.content).toBe("Adds 50% attack speed");
            done();
    })
    it("changes the benefit of the boon", async (done) => {
            const response = await request
                .patch("/boon/" + boonID)
                .send({content: "Resurrects fallen enemies"});
            expect(response.status).toBe(200);
            done();
    })
    it("gets the updated boon with the updated benefit", async(done) => {
            const response = await request
                .get("/boon/" + boonID)
            expect(response.status).toBe(200);
            expect(response.body.content).toBe("Resurrects fallen enemies");
            done();
    })
    it("deletes the ticket", async (done) => {
            const response = await request
                .delete("/boon")
                .send({uuid: boonID});
            expect(response.status).toBe(200);
            done();
    })
    it("will not find boon, since its deleted", async(done) => {
            const response = await request
                .get("/boon/" + boonID)
            expect(response.status).toBe(404);
            done();
    })
    it("deletes the god", async (done) => {
            const response = await request
                .delete("/god")
                .send({uuid: godID});
            expect(response.status).toBe(200);
            done();
    })
    it("will not find god, since its deleted", async(done) => {
            const response = await request
                .get("/god/" + godID)
            expect(response.status).toBe(404);
            done();
    })
});



hades-api/src/server.js
const express = require('express');
const bodyParser = require('body-parser');
const http = require('http');
const Helpers = require('./utils/helpers.js');
const port = 3000


const pg = require('knex')({
  client: 'pg',
  version: '9.6',      
  searchPath: ['knex', 'public'],
  connection: process.env.PG_CONNECTION_STRING ? process.env.PG_CONNECTION_STRING : 'postgres://example:example@localhost:5432/test'
});


const app = express();
http.Server(app); 


app.use(bodyParser.json());
app.use(
  bodyParser.urlencoded({
    extended: true
  })
);  

app.get('/test', (req, res) => {
  res.status(200).send();
});

/*******************************/
/*       Boon Endpoints        */
/*******************************/


/**
 * Post boon to boonTable
 * @params godname, uuid, content
 * @returns 
 */
app.post("/boon", (req, res) => {
  let uuid = Helpers.generateUUID();
    pg.insert({
      uuid: uuid,
      godname: req.body.godname,
      content: req.body.content,
      created_at: new Date(),
    })
      .into("boonTable")
      .then(() => {
        res.json({ uuid: uuid });
    });
});

/**
 * Get boon by UUID
 * @param uuid 
 * @returns  1 boon with uuid = req.params.uuid
 */
app.get("/boon/:uuid", async (req, res) => {
    const result = await pg.select(['uuid', 'godname', 'content', 'created_at']).from("boonTable").where({uuid: req.params.uuid});
    if(result.length == 0){
      res.status(404).send("Boon not found");
    } else{
      res.status(200).send(result[0]);
    }
});


/**
 * Get all boons in boonTable
 * @param
 * @returns All boons
 */
app.get('/boons', async (req, res) => {
  const result = await pg
    .select('*')
    .from('boonTable')
  res.json({
    res: result
  })
  res.status(200).send()
})


/**
 * Updates specific boon
 * @param uuid 
 * @returns
 */
app.patch("/boon/:uuid", async (req, res) => {
    pg('boonTable')
      .where({uuid: req.params.uuid})
      .update(req.body)
});


/**
 * Deletes specific boon
 * @param uuid 
 * @returns
 */
app.delete("/boon", (req, res) => {
    pg('boonTable')
      .where({ uuid: req.body.uuid })
      .del()
      .then(() => {
        res.sendStatus(200);
    })
});


/*******************************/
/*       Gods Endpoints        */
/*******************************/

/**
 * Gets a specific god by uuid
 * @param uuid 
 * @returns god with uuid
 */
app.get('/god/:uuid', async (req, res) => {
  const result = await pg.select(['uuid', 'name', 'description', 'created_at']).from('godTable').where({uuid: req.params.uuid})
  if(result.length == 0){
    res.status(404).send("Boon not found");
  } else{
    res.status(200).send(result[0]);
  }
});

/**
 * Post god to godTable
 * @params name, uuid, description
 * @returns
 */
app.post('/god', async (req, res) => {
  const uuid = Helpers.generateUUID();
  const result = await pg
    .insert({
      uuid,
      name: req.body.name,
      description: req.body.description,
      created_at: new Date(),
    })
    .table('godTable')
    .returning('*')
    .then((res) => {
      return res
    })
  res.status(200);
  res.json({
    res: result
  })
});

/**
 * Deletes specific god
 * @param uuid 
 * @returns
 */
app.delete("/god", (req, res) => {
    pg('godTable')
      .where({ uuid: req.body.uuid })
      .del()
      .then(() => {
        res.sendStatus(200);
    })
});

/*******************************/
/*         Table inits         */
/*******************************/

async function initialiseTables() {
  await pg.schema.hasTable('godTable').then(async (exists) => {
    if (!exists) {
      await pg.schema
        .createTable('godTable', (table) => {
          table.increments();
          table.uuid('uuid');
          table.string('name');
          table.string('description');
          table.timestamps(true, true);
        })
        .then(async () => {
          console.log('created table god');
          for (let i = 0; i < 10; i++) {
            const uuid = Helpers.generateUUID();
            await pg.table('godTable').insert({ 
              uuid, 
              name: `god number ${i}`,
              description: 'A god of Olympus'
            })
          }
        });
        
    }
  });

  await pg.schema.hasTable('boonTable').then(async (exists) => {
    if (!exists) {
      await pg.schema
        .createTable('boonTable', (table) => {
          table.increments();
          table.uuid('uuid');
          table.string('godname');
          table.string('content');
          table.timestamps(true, true);
        })
        .then(async () => {
          console.log('created table boon');
        });
    }
  });
  
}
initialiseTables()

module.exports = app;